src/storage/src/upsert/types.rs

```diff
@@ -623,16 +680,72 @@ where
     where
         G: IntoIterator<Item = UpsertKey>,
         R: IntoIterator<Item = &'r mut UpsertValueAndSize<O>>;
+
+    /// For each key in `merges` writes a 'merge operand' to the backend. The backend stores these
+    /// merge operands and periodically calls the `snapshot_merge_function` to merge them into
+    /// any existing value for each key. The backend will merge the merge operands in the order
+    /// they are provided, and the merge function will always be run for a given key when a `get`
+    /// operation is performed on that key, or when the backend decides to run the merge based
+    /// on its own internal logic.
+    /// This allows avoiding the read-modify-write method of updating many values to
+    /// improve performance.
+    ///
+    /// The `MergeValue` should include a `diff` field that represents the update diff for the
+    /// value. This is used to estimate the overall size diff of the working set
+    /// after the merge operands are merged by the backend `sum[merges: m](m.diff * m.size)`.
+    ///
+    /// `MergeStats` **must** be populated correctly, according to these semantics:
+    /// - `written_merge_operands` must record the number of merge operands written to the backend.
+    /// - `size_written` must record the total size of values written to the backend.
+    ///     Note that the size of the post-merge values are not known, so this is the size of the
+    ///     values written to the backend as merge operands.
+    /// - `size_diff` must record the estimated diff of the total size of the working set after the
+    ///    merge operands are merged by the backend.
+    async fn multi_merge<P>(&mut self, merges: P) -> Result<MergeStats, anyhow::Error>
+    where
+        P: IntoIterator<Item = (UpsertKey, MergeValue<StateValue<O>>)>;
+}
+
+/// A function that merges a set of updates for a key into the existing value for the key, expected
+/// to only be used during the snapshotting-phase of an upsert operator. This is called by the
+/// backend implementation when it has accumulated a set of updates for a key, and needs to merge
+/// them into the existing value for the key.
+///
+/// The function is called with the following arguments:
+/// - The key for which the merge is being performed.
+/// - An iterator over any current value and merge operands queued for the key.
+/// The function should return the new value for the key after merging all the updates.
+pub(crate) fn snapshot_merge_function<O>(
+    _key: UpsertKey,
+    updates: impl Iterator<Item = StateValue<O>>,
```

Comment:
the real perf question is how expensive it is to decode all these `StateValue`'s into `value_xor` `Vec`'s 

worth a `TODO` to supporting `Cow<[u8]>` in `StateValue` and seeing if we can use `serde` borrow support to avoid these extra allocations: https://serde.rs/lifetimes.html#borrowing-data-in-a-derived-impl; that way each merge function instantiation costs a singular allocation (with some possible resizes)

Reply:
Oh row doesn't support this (yet), ill talk to parker!

Reply:
Actually, we should 100% wrap the `Vec<u8>` buffer in `Snapshotting` in `https://docs.rs/serde_bytes/latest/serde_bytes/index.html`, thats a good start to reduce some cpu cycles

Reply:
@guswynn should I tackle that in this PR, or are you suggesting as a future optimization?

Reply:
I think its fine as a future optimization; if you see bad perf/high cpu usage in your testing its probably worth it though!

Reply:
I'm not seeing an improvement in my perf testing so this seems like something we should do now, though I'm going to merge this PR as-is so the next review is a clean slate based around perf improvements, and this sits behind a feature flag for now.

Reply:
sounds good!